snapcraft/plugins/jhbuild.py

```diff
@@ -0,0 +1,383 @@
+                'default': 'gnome-world',
+            },
+            'module-set-dir': {
+                'type': 'string',
+                'default': '',
+            },
+            'jhbuild-archive': {
+                'type': 'string',
+                'default': '',
+            },
+            'jhbuild-mirror': {
+                'type': 'string',
+                'default': '',
+            },
+            'cflags': {
+                'type': 'string',
+                'default': '',
+            },
+        }
+
+        schema['required'] = ['modules']
+
+        return schema
+
+    @classmethod
+    def get_pull_properties(cls):
+        # Inform Snapcraft of the properties associated with pulling. If these
+        # change in the YAML Snapcraft will consider the build step dirty.
+        return [
+            'modules',
+            'module-set',
+            'module-set-dir',
+            'jhbuild-archive',
+            'jhbuild-mirror',
+        ]
+
+    @classmethod
+    def get_build_properties(cls):
+        # Inform Snapcraft of the properties associated with building. If these
+        # change in the YAML Snapcraft will consider the build step dirty.
+        return [
+            'cflags',
+        ]
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self.modules = [
+            module for module in self.options.modules if module[0] != '-']
+
+        self.skip_modules = [
+            module[1:] for module in self.options.modules if module[0] == '-']
+
+        self.build_packages += list(BUILD_PACKAGES)
+
+        self.jhbuild_src = os.path.join(self.partdir, 'jhbuild', 'src')
+        self.jhbuild_program = os.path.join(
+            self.partdir, 'jhbuild', 'usr', 'bin', 'jhbuild')
+        self.jhbuildrc_path = os.path.join(self.partdir, 'jhbuildrc')
+
+    def enable_cross_compilation(self):
+        """
+        We don't support cross compilation (yet?)
+        """
+        raise NotImplementedError(
+            'The {!s} plugin does not support cross compilation'.format(
+                self.name))
+
+    def run(self, cmd, cwd=None, **kwargs):
+        """Run a command.
+
+        Use run_output() if you need to capture or suppress output.
+
+        :param list cmd: command arguments, first is the executable
+        :param str cwd: working directory
+        """
+        return super().run(cmd, cwd=cwd, **kwargs)
+
+    def run_output(self, cmd, cwd=None, **kwargs):
+        """Run a command, capturing its output.
+
+        Use run() if you don't need the output and prefer it to be piped to
+        stdout.
+
+        :param list cmd: command arguments, first is the executable
+        :param str cwd: working directory
+        :return: the output of the command
+        :rtype: str
+        """
+        return super().run_output(cmd, cwd=cwd, **kwargs)
+
+    def jhbuild(self, args, output=True, **kwargs):
+        """Run JHBuild in the build stage.
```

Comment:
Indeed, "stage" is actually a step :stuck_out_tongue: . Terminology minefield here.